libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -379,17 +379,21 @@ void NameAndTypeResolver::linearizeBaseContracts(ContractDefinition& _contract)
 		// "push_front" has the effect that bases mentioned later can overwrite members of bases
 		// mentioned earlier
 		input.back().push_front(base);
+		_contract.annotation().contractDependencies.emplace(base, baseSpecifier.get());
```

Comment:
Do we actually still need this, or are the dependencies from the call graph enough?

Reply:
If we have
```
contract A { function f() public { new B(); } }
contract B is A {}
```
this will be caught through the call graph.
A special case that changes is if the base function is not executed when the derived is the "most derived":
```
contract A { function f() public virtual { new B(); } }
contract B is A {
 function f() public override {  }
}
```
But in that case, it is actually possible for `A` to create `B`.

Reply:
But I think there are some cases where adding stuff to `contractDependencies` is now redundant. For example https://github.com/ethereum/solidity/pull/10228#discussion_r587791629.

Though maybe not. The call graph does not take into account stuff that is unreachable. For example if there's a creation cycle caused by unused internal functions, it won't be detected.
```solidity
contract C { function foo() internal { new D(); } }
contract D { function foo() internal { new C(); } }
```
Currently it causes an error:
```
Error: Circular reference for contract creation (cannot create instance of derived or same contract).
```
But the graph won't report `C` and `D` in `createdContracts`.

Reply:
What do you mean by "used in member access expressions"?
The reason CompilerStack::compileContract uses `contractDependencies` is because `new C` needs to have `C` available, and that's the only reason.

Reply:
I meant that the `TypeChecker` was adding contracts used in member access expressions to `contractDependencies` and function call graph does not add them to `createdContracts` so it could not be used as a replacement.

But well, I see your point now. It should be possible to generalize `createdContracts` to just store info about dependencies rather than just about `new` expressions.

Reply:
I ran a test and if we don't add them here, our tests all still work, EXCEPT the astjson tests who know are missing their contract dependency information for their field `contractDependencies`. I'll look into the callgraph and check if we can make it do that somehow..

Reply:
Alright, I misinterpreted a previous error wrongly and it seems this does indeed break tests involving libraries like this one:

```
library L {
    function f() internal {
        new C();
    }
}

contract D {
    function f() public {
        L.f();
    }
}
contract C {
    constructor() { new D(); }
}
```

The problem is that we never detect the `new C();` in `L.f()` and never fill the dependency `D` for `L`..

Reply:
So the call graph does not properly handle internal library functions? I think there should not be a dependency from L to C but rather from D to C, because the code of L.f is included directly in D.

Reply:
That would also explain the problem I am seeing. I'll see if I can fix my problem with that information..